### PR TITLE
Add `doomPackageDir` argument.

### LINKED ISSUE #297: Build native-compiles every package every time my emacs config changes.
Hi!

I use `nix-doom-emacs` [like this](https://github.com/znewman01/dotfiles/blob/98366c4aab71385c328ab28a3bbaabb3377e4bf0/emacs/default.nix#L6-L12), following the suggestion in https://github.com/nix-community/nix-doom-emacs/issues/60#issuecomment-1083630633 for tangling my literate config ([here](https://github.com/znewman01/dotfiles/blob/9174aafb6bd4d03aa8e31b7348e572cb6567a052/emacs/doom.d/default.nix)).

```nix
  programs.doom-emacs = {
    enable = true;
    doomPrivateDir = (import ./doom.d) {
      inherit lib;
      inherit (pkgs) stdenv emacs coreutils;
    };
  };
```

Every time I change my Emacs configuration at all and rebuild, it takes a *long* time while Nix recompiles every Emacs package. I have the [nix-community cachix added](https://github.com/znewman01/dotfiles/blob/98366c4aab71385c328ab28a3bbaabb3377e4bf0/common/default.nix#L5-L9), but it doesn't appear to get used at all. Not sure why.

I want to figure out the problem with cachix but in the meantime, I noticed that this rebuild happens even if the set of packages changed. I played around with it, and turns out it's pretty easy to cache the straight-env by only including `init.el` and `packages.el`: https://github.com/znewman01/nix-doom-emacs/commit/a076453f02efa5016511a0154423afd711f81d19

Would something like that be useful to commit back? It feels strictly better to me than rebuilding each time. Maybe a flag to turn off the behavior in case someone has an interesting package installation setup that relies on some other files?

For reference: my [flake.nix](https://github.com/znewman01/dotfiles/blob/98366c4aab71385c328ab28a3bbaabb3377e4bf0/flake.nix#L18-L21) (pointing at my fork of nix-doom-emacs) and [flake.lock](https://github.com/znewman01/dotfiles/blob/98366c4aab71385c328ab28a3bbaabb3377e4bf0/flake.lock#L24).

### PR DESCRIPTION
**Pull request by znewman01**
This is a directory used for initializing the straight environment, distinct from `doomPrivateDir`. If not given, `doomPrivateDir` is used.

Here it is in action:

https://github.com/znewman01/dotfiles/blob/be9f3a24c517a4ff345f213bf1cf7633713c9278/emacs/default.nix#L12-L34

Fixes #297 (there are some ideas for extra sugar in that issue, but this solves the core problem).

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -1,5 +1,23 @@
 { # The files would be going to ~/.config/doom (~/.doom.d)
```
Comment (ckiee) [inline default.nix:1]: I think a check in the flake.nix could be nice; `init-example-el-splitdir` or something.

File: default.nix [hunk 2]
```diff
@@ -1,5 +1,23 @@
...
+     Example:
+       doomPackageDir = pkgs.linkFarm "my-doom-packages" [
+         # straight needs a (possibly empty) `config.el` file to build
+         { name = "config.el"; path = pkgs.emptyFile; }
+         { name = "init.el"; path = ./doom.d/init.el; }
+         { name = "packages.el"; path = pkgs.writeText "(package! inheritenv)"; }
+         { name = "modules"; path = ./my-doom-module; }
+       ];
+   */
+,  doomPackageDir ? null
```
Comment (thiagokokada) [inline default.nix:20]: Actually this could also be expressed as:

```suggestion
,  doomPackageDir ? doomPrivateDir
```

File: default.nix (hunk 1 above)
Comment (thiagokokada) [reply default.nix:1]: Yeah, I coucur. Would be nice to add a test for this case.

Comment (znewman01) [reply default.nix:1]: Good idea! Check the test I just added.

File: default.nix (hunk 2 above)
Comment (znewman01) [reply default.nix]: Great call, done.

File: checks.nix [hunk 3]
```diff
@@ -41,4 +41,26 @@ in
...
+    dependencyOverrides = inputs;
+    doomPrivateDir = pkgs.linkFarm "my-doom-packages" [
+         { name = "config.el"; path = ./test/doom.d/config.el; }
+         { name = "init.el"; path = ./test/doom.d/init.el; }
+         # Should *not* fail because we're building our straight environment
+         # using the doomPackageDir, not the doomPrivateDir.
+         {
+           name = "packages.el";
+           path = pkgs.writeText "packages.el" "(package! not-a-valid-package)";
+         }
```
Comment (thiagokokada) [inline checks.nix:54]: Good test :+1: .

File: modules/home-manager.nix [hunk 4]
```diff
@@ -22,6 +22,27 @@ in
...
+
+        Can be used, for instance, to prevent rebuilding the Emacs environment
+        each time the `config.el` changes.
+
+        Can be provided as a directory or derivation. If not given, package
+        environment is built against `doomPrivateDir`.
+      '';
+      apply = path: if lib.isStorePath path then path else builtins.path { inherit path; };
+      example = literalExample ''
+       doomPackageDir = pkgs.linkFarm "my-doom-packages" [
```
Comment (thiagokokada) [inline modules/home-manager.nix:37]:
```suggestion
        doomPackageDir = pkgs.linkFarm "my-doom-packages" [
```

**Review by thiagokokada — COMMENTED** [review]
Minor nit.

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
Merging it.

We probably want to add this to #266.

**Comment by quinn-dougherty** [comment]
The syntax for this isn't ideal. Hiding `config.el` in the private dir and replacing it with the empty file in the package dir would ideally be replaced with something that hides all that from the caller. This was my first crack at it https://github.com/quinn-dougherty/configuration.nix/blob/37c6e932c26a7d11efd9e39816bb9966d7445b71/framework/users/qd/packages/emacs.nix tho I'm concerned I didn't get something right.

**Pull request merged**